disvae/losses.py

```diff
@@ -204,27 +212,27 @@ def __call__(self, data, model, optimizer, is_train, storer):
         batch_size = data.size(dim=0)
         half_batch_size = batch_size // 2
         data = data.split(half_batch_size)
-        data1 = data[0].to(self.device)
-        data2 = data[1].to(self.device)
+        data1 = data[0]
+        data2 = data[1]
 
         # Initialise the targets for cross_entropy loss
-        ones = torch.ones(batch_size, dtype=torch.long, device=self.device)
-        zeros = torch.zeros(batch_size, dtype=torch.long, device=self.device)
+        ones = torch.ones(half_batch_size, dtype=torch.long, device=self.device)
+        zeros = torch.zeros(half_batch_size, dtype=torch.long, device=self.device)
 
         # Get first sample of latent distribution
-        recon_batch, latent_dist = model(data1)
+        recon_batch, latent_dist, latent_sample = model(data1)
 
         # Get reconstruction loss
         rec_loss = _reconstruction_loss(data1, recon_batch, self.is_color)
 
         # Get KL-Divergence (latent_dist[0] = mean, latent_dist[1] = log_var)
         kl_loss = _kl_normal_loss(*latent_dist, storer)
 
-        # Run latent distribution through discriminator
-        d_z = self.discriminator(latent_dist)
+        # Run latent sample through discriminator
+        d_z = self.discriminator(latent_sample)
 
         # Calculate the total correlation (TC) loss term
-        tc_loss = (d_z[:, :, :1] - d_z[:, :, 1:]).mean()
+        tc_loss = (d_z[:, :1] - d_z[:, 1:]).mean()
```

Comment:
so decided to keep 2 outputs? Maybe for interest of time we can change/test that after the poster

Reply:
I made a comment on the original pull request but I don't think you saw it, so repeated here:

I took another look at the paper and they say in Appendix A:

> We use the same 6 layer MLP discriminator with 1000 hidden units per layer and leaky ReLU (lReLU) non-linearity, that outputs 2 logits in all FactorVAE experiments.

So I think the discriminator implementation is actually correct. Is the constraint enforced by this line of code?:

`d_tc_loss = 0.5 * (F.cross_entropy(d_z, zeros) + F.cross_entropy(d_z_perm, ones))`

Reply:
Agreed, we can explore this more after the poster.